### PR TITLE
JIT: Add a uniform representation for parameter ABI information

### PR DESCRIPTION
**Pull request by jakobbotsch**
This adds a uniform representation that can represent the ABI
information for all of our targets without needing to fall back to
handling ABI specific details in all places that need to handle calling
conventions.

Currently nothing is using this information. I want to incrementally
migrate our ABI handling to use this representation. Also, there are
several potential future improvements:

- Split out ABI classification per ABI instead of keeping them all
  within the same function
- Unify `InitVarDscInfo::stackArgSize` and `InitVarDscInfo::argSize`. I
  am unsure why the latter is needed
- Remove `LclVarDsc::GetArgReg()`, `LclVarDscInfo::GetOtherArgReg()`,
  HFA related members
- Reuse the representation in `CallArgABIInformation` and unify the
  classifiers

The end goal here is rewriting `genFnPrologCalleeRegArgs` to handle
float and integer registers simultaneously, and to support some of the
registers that the Swift calling convention is using.

### CODE REVIEW

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @AndyAyersMS 

[No diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=615681&view=ms.vss-build-web.run-extensions-tab). Some minor TP regressions (the seemingly large MinOpts ones are in collections with < 10 MinOpts contexts).

There's a lot more clean up to be done based on this, but I initially want to use this to represent the ABI details for the structs in Swift reverse pinvokes.

File: src/coreclr/jit/abi.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#pragma once
+
+class ABIPassingSegment
+{
+    regNumber m_register    = REG_NA;
+    unsigned  m_stackOffset = 0;
+
+public:
+    bool IsPassedInRegister() const;
+    bool IsPassedOnStack() const;
```
Comment (tannergooding) [inline src/coreclr/jit/abi.h:13]: Is it relevant to track the other common qualifications like `HFA` and `HVA`?

Comment (jakobbotsch) [reply src/coreclr/jit/abi.h:13]: I don't think so. HFA's/HVA's are just passed in more registers than other struct arguments, so hopefully by the end of this clean up that's a detail that's constrained fully to be within the ABI classification and not leaked out anywhere to the rest of the JIT.

File: src/coreclr/jit/abi.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+struct ABIPassingInformation
+{
+    // The number of segments used to pass the value. Examples:
+    // - On x86, TYP_LONG can be passed in two registers, resulting in two
+    // register segments
+    // - On SysV x64, structs can be passed in two registers, resulting in two
+    // register segments
+    // - On arm64/arm32, HFAs can be passed in up to four registers, giving
+    // four register segments
```
Comment (tannergooding) [inline src/coreclr/jit/abi.h:44]: This should also include HVAs, which looks like we might be missing handling around:

[godbolt HFA/HVA for Arm64](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIAOxcpK4AMngMmAByPgBGmMQSAMykAA6oCoRODB7evgFBaRmOAmER0SxxCVzJdpgOWUIETMQEOT5%2BgbaY9sUMjc0EpVGx8Um2TS1teZ0KE4PhwxWj1QCUtqhexMjsHOaJ4cjeWADUJoluzSwA%2BhECAHQIZ9gmGgCCL6%2BzxF4OxwASADFXqd/BZjlRaKgmARjqpSMcAJ7wgBe8IA7mcwSZ/AARTEfL4/GF/ABqwOxYIhUIIiTMqkkVxhcMRKPRmJBeMSVjeHyp0OOAFkIIDgQgqEwVh8KR9jrLjsRMARNgxjmKmHdVKdLKrxXcEVqwWq7siDTr1RiuVLcVK3nyaXSGTDIsKyaqAG4Sq3c4FyhVK4gqj3odAARyuVFpECDofDkejYYjZggCA9GvhKfVCLW7vVyOzGbuaMllp5uI4a1onAArLw/BwtKRUJw3NZrMcFBstpgtYkeKQCJpy2sANYgKtcO5mLgANmn/ir064ZgAHFwNMuAJz6TiSWuDxucXgKEAafuDtZwWAwRAoVAsFJ0eLkShoe%2BPhKXN3LgC0h0MwG/AhvgYYc%2BDoAh4mPCAYn3GJwmaBFOD7ODmGIBEAHkYm0OoB24XhXzYQR0IYWhEPrXgsBiLxgAuWhaGPPDSCwFh/3EcimLwBV6jdTAGIbTBVDqLwIKQ3hwggyt2NoPAYmIBCPCwfcgLwFhRLWCEmGABQSTwTA0XQlJGFEmRBBEMR2CkEz5CUNR910IIDCMFBW0sfQZOPSA1lQFJegY790MSXhUB44hiDwLAPIgNZanqZwIFcKY/CCUIFnKSo9EKTIBESjL0iyhghjS5Yuh6Bo5hyoIYt6foWkKkYqnGAYKsa2rUvqiRos7bYOu3Dga1IOsGybDhjk/H8mGOP8jGOCAgK8ECVhm3BCBIHsuBWXhcK0FYRxASRlzufxl0SWlJESZdAk3Jdp163dSFUqtT0GoLD1sE8z3Ii9r2TW833oMgKAgV8H3%2BkA3WQFIUiuN0uA3K4DAg2YrkuadpBoWgIOIKCYPYlCEOMvG0Mw7CHGMgjGAIYjSP3SjqNo%2BjjOY1idgbfAuMcHi%2BN4AShJExjxO6fdpNk%2BSMBZzawtUvD1IMLSdL0gyjMY/hTNEcRLJV6yVHUdjdDMfR/2cyxrDcmJIq8nysj89CzCCkKwoi%2BBou6HCshcBh3E8do9BSsp2oKPLemazLejqpYGqqsqmq9vJKpd2K%2BjmMP0sq8qY6Slr5j98Oeo7TZuvW3r%2Bueg8RpRyRjhYBRweOGGNzuBHeJhCBlqIYg1o2j7trWBBMCYLAEii27eAep792Go93q2odSFHfbDuO07zsujdrt6wKBvH17p523rbc39iJ67meQoyZxJCAA)

[godbolt HVA/HVA for x64](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:67,endLineNumber:13,positionColumn:67,positionLineNumber:13,selectionStartColumn:67,selectionStartLineNumber:13,startColumn:67,startLineNumber:13),source:'%23include+%3Cxmmintrin.h%3E%0A%0Astruct+HFA+%7B+float+x,+y,+z,+w%3B+%7D%3B%0Astruct+HVA+%7B+__m128+x,+y,+z,+w%3B+%7D%3B%0A%0Afloat+M(HFA+hfa)%0A%7B%0A++++return+hfa.x+%2B+hfa.y+%2B+hfa.z+%2B+hfa.w%3B%0A%7D%0A%0A__m128+N(HVA+hva)%0A%7B%0A++++return+_mm_add_ps(_mm_add_ps(_mm_add_ps(hva.x,+hva.y),+hva.z),+hva.w)%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:51.21667561281089,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:clang_trunk,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-O3',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+(trunk)+(Editor+%231)',t:'0')),k:48.783324387189126,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:vcpp_v19_latest_x64,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:2,lang:c%2B%2B,libs:!(),options:'-O2',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x64+msvc+v19.latest+(Editor+%231)',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:48.783324387189126,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

Comment (jakobbotsch) [reply src/coreclr/jit/abi.h:44]: These are just examples to indicate examples of the representation. The intention of this PR is not to make any functional changes.

**Review by AndyAyersMS — APPROVED** [review]
Seems like a good start...

**Pull request merged**